Handle errored loading states better

rtibbles opened the pull request:
## Summary
* Rehomes KContentRenderer in Kolibri from KDS
* Simplifies its implementation which had been made more complex for the abstraction into KDS
* Set language globals in the module scope, as there's no longer a need to defer them, and it could negatively affect the languageDirection used in the pluginMediator
* Updates ContentPage.vue to stop tracking progress when content renderer errors
* Fixes long standing bug where activity in renderers was not properly toggling on learner activity for the session
* Updates the zip content backend to return HTML pages with embedded meta to allow reading for error reports once the iframe has loaded
* Defines a LOADING_ERROR constant and uses it to emit loading errors from all the content renderers
* Handles these loading errors in TopicsContentPage when there is a defined deviceId and shows a modal inviting the user to return to all libraries
* Adds additional error alerts with a 'report this problem' pop up to allow users to more easily report errors with specific resources
* Adds a short expiry to the locks used for reserving downloads in the ChunkedFile implementation to ensure that range requests for videos don't get permanently stalled (final resolution of video streaming to follow in subsequent PRs).

## References
Fixes [#10393](https://github.com/learningequality/kolibri/issues/10393)
Fixes [#10603](https://github.com/learningequality/kolibri/issues/10603)
Fixes https://github.com/learningequality/kolibri/issues/10124

## Reviewer guidance
Browse HTML5 apps and Videos on remote Kolibris to confirm that the error modal properly displays.

![Screenshot from 2023-05-24 11-33-04](https://github.com/learningequality/kolibri/assets/1680573/740ff429-1467-4ef7-a3f3-471e5bf4dcaa)


----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [x] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

rtibbles committed:
Reinstate KContentRenderer into the core Kolibri codebase.

rtibbles committed:
Set language globals as soon as possible.

rtibbles committed:
Simplify content rendering to remove abstraction introduced for externalization.

rtibbles committed:
Turn KContentRenderer into a functional component.
Move listener adding into content mixin.

rtibbles committed:
Update tests.

rtibbles committed:
Stop tracking progress when content errors.

rtibbles committed:
Add specific loading errors for content and handle them in the topics content page
in the case of remote browsing.

rtibbles committed:
Fix content renderer mixin mixing.

rtibbles committed:
Tweak mixin to add events and register event handling better.
Use constant for RTL.

rtibbles committed:
Make report error modal more reusable.

rtibbles committed:
Display error alert when renderer triggers error event.

rtibbles committed:
Emit error messages from hashi when the backend returns an explicit error page.

rtibbles committed:
Cleanup script loader imports.

rtibbles committed:
Add loading error handling to all other renderers.

rtibbles committed:
Add an expiry to the download Lock to allow reacquisition after cancel.

rtibbles committed:
Ensure parity of event emission from different components.

nucleogenesis commented on kolibri/core/assets/src/views/ContentRenderer/ContentRendererError.vue:
```diff
@@ -1,21 +1,52 @@
-    {{ $tr('rendererNotAvailable') }}
-  </UiAlert>
+  <div>
+    <UiAlert :dismissible="false" class="alert" type="error">
```
This gets me wondering if I should use [`UiAlert`](https://josephuspaye.github.io/Keen-UI/#/ui-alert) in #10739 (and then also for the `StorageNotificationBanner`) rather than applying styling ourselves?

Alternatively, perhaps we should just wrap and extend `UiAlert` in KDS as a "Banner" component.

Just a couple thoughts probably worth chatting through in an upcoming KDS Tactical

nucleogenesis replied:
The UiAlert is getting mushed w/ the Video player and it looks similar on the PDF renderer so might need some margin there.

![image](https://github.com/learningequality/kolibri/assets/6356129/7b280dbd-ea02-4a8d-8897-08b127cdba39)

rtibbles replied:
I did deliberately set the margin-bottom to 0 to make it flush to the renderer below, but seems that might not be desirable.

rtibbles replied:
I consulted with @jtamiace and have now updated this to an 8px margin all around the UIAlert. To ensure that it showed consistently, I also overrode the default UIAlert width to account for the extra margin on either side.

nucleogenesis commented on kolibri/plugins/perseus_viewer/assets/src/views/PerseusRendererIndex.vue:
```diff
@@ -91,7 +91,7 @@
   import urls from 'kolibri.urls';
   import responsiveWindowMixin from 'kolibri.coreVue.mixins.responsiveWindowMixin';
-  import scriptLoader from 'kolibri.utils.scriptLoader';
+  import scriptLoader from 'kolibri-common//utils/scriptLoader';
```
`kolibri-common//` are the double slashes intended? Any chance this could cause some kind of issue?

rtibbles replied:
Apparently my right pinky got carried away. Weirdly, I think this isn't causing an issue, but worth cleaning up regardless.

rtibbles replied:
Fixed!

nucleogenesis commented on kolibri/plugins/learn/assets/src/views/TopicsContentPage.vue:
```diff
@@ -134,7 +135,16 @@
+      v-if="showConnectionErrorModal"
+      :title="deviceFormTranslator.$tr('errorCouldNotConnect')"
+      :submitText="coreString('goBackAction')"
+      @submit="goToAllLibraries"
```
Is there a case where one piece of content doesn't work but others might from this remote device -- or is this an indication that the remote device cannot serve content to you right now in general?

rtibbles replied:
If Kolibri is operating correctly, it should be the latter. At the moment, this is not the case, but I'll be hopefully fixing that in other PRs!

nucleogenesis reviewed (dismissed):
I triggered the error by blocking the URL in dev-tools for the media request and got the modal as expected. Just have a few style notes re: the ContentError component. 

Just a minor style note on the change request front.

Also -- I had to grep Studio because I was convinced we used KContentRenderer there. I feel like there was some good reason and plan for it living in KDS but it's lost to the ether.

---

Before this merges think I should pull and link your corresponding KDS draft PR and make sure the transplant is as seamless as it appears?

rtibbles commented:
> Also -- I had to grep Studio because I was convinced we used KContentRenderer there. I feel like there was some good reason and plan for it living in KDS but it's lost to the ether.

The original plan for moving it first into `kolibri-components` and then into KDS was to make KContentRenderer a truly portable component. However, there's no good way to use the renderer plugins outside of Kolibri itself, so it ended up not being a helpful migration.

Keeping it a core part of Kolibri seems like the best path forward, and instead, we will turn Studio into a Kolibri plugin in order to be able to leverage the renderer.

rtibbles commented:
> Before this merges think I should pull and link your corresponding KDS draft PR and make sure the transplant is as seamless as it appears?

I think that would be helpful, yes. If it is smooth, we can merge the KDS PR and I can do the KDS upgrade within this PR.

rtibbles committed:
Update margin to 8px all round.
Fix typo in module path.

marcellamaki commented:
> we will turn Studio into a Kolibri plugin

😄

marcellamaki reviewed (changes requested):
The changes within the content renderer seem to be handled well, but one scenario I encountered while testing this is going back to the topics page if the disconnect happens during content but before the popup happens, or after the content is completed. 

I just see a blank screen and get a 410 when going back without the modal prompt

Here is an example where much of the video has loaded, so though the "disconnected" state displays, the buffered video is able to continue. But if I navigate out, I am not able to go back.

https://github.com/learningequality/kolibri/assets/17235236/d94b1bf6-2266-4c5f-94f6-09af600bd251

rtibbles committed:
Redirect to the library page if API requests error with 410.

rtibbles commented:
@marcellamaki I have added some additional handling for the 410 status code specifically, which will redirect to the Library page in the edge case you have described. It would probably be better to display some sort of overlay alert or somesuch to the user after this has happened, but that may be something for a future patch (and will need a little design).

rtibbles committed:
Remove defunct cache busting for API calls and rely on cache headers.

rtibbles committed:
Move content cache key handling out of api-resource and into specific resources.

rtibbles committed:
Don't use content cache key for remote requests.

rtibbles committed:
Don't proxy content-type header.
Update caching to always set a cache timeout but conditionalized on content cache key.

rtibbles committed:
Catch failed search labels fetching.

rtibbles commented:
Added some additional handling here to prevent error requests being cached for a long time. Allowed successful requests to be cached for a short time regardless of whether the content cache key is set. I also added a simple error catch for updating the search labels. I hope this should resolve the remaining issues. I decided against shortening the request timeout, as I think that would be better handled by showing appropriate loading states.

marcellamaki reviewed (approved):
After chatting a bit more with @rtibbles - I feel good to approve this. 

Additional UI updates around loading state and UI messaging can be handled in follow up, and the error handling itself is managed here, so good to merge.
